handler/src/main/java/io/netty/handler/ssl/OpenSslContext.java

```diff
@@ -226,13 +229,20 @@ public final SSLEngine newEngine(ByteBufAllocator alloc, String peerHost, int pe
     @Override
     public final SSLEngine newEngine(ByteBufAllocator alloc) {
         List<String> protos = applicationProtocolNegotiator().protocols();
+        OpenSslEngineMap engineMap = engineMap();
+        final OpenSslEngine engine;
         if (protos.isEmpty()) {
-            return new OpenSslEngine(ctx, alloc, null, isClient(), sessionContext());
+            engine = new OpenSslEngine(ctx, alloc, null, isClient(), sessionContext(), engineMap);
         } else {
-            return new OpenSslEngine(ctx, alloc, protos.get(protos.size() - 1), isClient(), sessionContext());
+            engine = new OpenSslEngine(ctx, alloc, protos.get(protos.size() - 1), isClient(),
+                    sessionContext(), engineMap);
         }
+        engineMap.add(engine);
+        return engine;
     }
 
+    abstract OpenSslEngineMap engineMap();
```

Comment:
I mainly did this as this depends on what X509TrustManager is used that is decided in the `OpenSslClientContext` and `OpenSslServerContext` constructors itself.